### ISSUE TITLE
Maximum Run Time Execution

### ISSUE THREAD

**Issue by isaaclo97** [rmartinsanta/mork#80]
Allow to configure a maximum run time execution per instnace in a application.yml property.
Some articles report fixed times (60 seconds, 3600 seconds), this parameter can allow us to select a maximum execution time to obtain same or less time depending on the purpose (despite being a different CPU/GPU).

**Comment by rmartinsanta** [comment]
After thinking about this, I do not see a way to implement it without user cooperation. I think it is always mandatory that the user takes into account time limits while programming its approaches, specially if they make custom Improvers, where most time will be spent. I can think of two approaches:

A) Global flag
- The framework measures time according to different criteria (reference time from other papers, fixed value, scaled due to benchmark, etc), and triggers the flag when the allocated resources have been exceeded.
- The user would be responsible for checking this flag frequently, specially in iterated approaches, so if the flag is triggered the execution is short circuited and the current best solution returned.

B) Using annotations
I may be able to create a custom annotation, such as `@TimeLimited`. The user can use this annotation to automatically stop execution of certain methods. Internally, it would work as follows: before a method annotated with `@TimeLimited` is executed, the time spent is checked, and if have already reached the limit, the method is NOT executed, and the return value (must be either void or a solution) is overriden to return the current solution (which should be a parameter of the same method). Example:

```
/**
     * {@inheritDoc}
     *
     * Improves a model.Solution
     * Iterates until we run out of time, or we cannot improve the current es.urjc.etsii.grafo.solution any further
     */
    @Override
    protected S _improve(S s) {
        S newSolution = iteration(s);
        while (newSolution.isBetterThan(s)){
           s = newSolution;
           newSolution = iteration(s);
        }
        return s;
    }

    /**
     * Tries to improve the recieved solution once
     *
     * @param s Solution to improve
     * @return True if the es.urjc.etsii.grafo.solution has been improved, false otherwise
     */
    @TimeLimited
    public abstract S iteration(S s);
    // Automatic time check each time the method is invoked, if the time limit is reached, parameter is inmediatly returned without executing the method.
```

**Comment by ea2809** [comment]
Sir @rmartinsanta, the user should always do a good job. By default, it could be done in the internal algorithms, and the documentation should explain how it works.

**Comment by rmartinsanta** [comment]
First option implemented in #95, second one may be added in the future.
Using it should be simple:
- Inside any algorithm component, `TimeControl.isTimeUp()` returns true if the algorithm should finish as soon as possible, false otherwise. 
- Disabled by default, is automatically enabled if the user creates a class extending `TimeLimitCalculator`, example:
```java
public class TimeLimitCalculatorExample extends TimeLimitCalculator<MySolution, MyInstance> {

    @Override
    public long timeLimitInMillis(MyInstance instance, Algorithm<MySolution, MyInstance> algorithm) {
        // 10 minutes execution time for any instance / algorithm combination
        // TimeLimit can be different per instance or algorithm, using any strategy that the user decides
        return 10 * 60 * 1000;
    }
}
```

Any feedback is welcome!

**Comment by rmartinsanta** [comment]
@ea2809 @isaaclo97 FYI it will be merged with autoconfig in next release v0.15, as it currently stands unless you send any additional feedback

**Comment by rmartinsanta** [comment]
Closing as completed, if there is any issue feel free to open a new issue.
